packages/ckeditor5-list/src/liststylesediting.js

```diff
@@ -0,0 +1,511 @@
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module list/liststylesediting
+ */
+
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
+import ListEditing from './listediting';
+import ListStylesCommand from './liststylescommand';
+import { getSiblingListItem } from './utils';
+
+const DEFAULT_LIST_TYPE = 'default';
+
+/**
+ * The list styles engine feature.
+ *
+ * It sets value for the `listItem` attribute for the {@link module:list/list~List `<listItem>`} element that
+ * allows modifying list style type.
+ *
+ * @extends module:core/plugin~Plugin
+ */
+export default class ListStylesEditing extends Plugin {
+	/**
+	 * @inheritDoc
+	 */
+	static get requires() {
+		return [ ListEditing ];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	static get pluginName() {
+		return 'ListStylesEditing';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	init() {
+		const editor = this.editor;
+		const model = editor.model;
+
+		// Extend schema.
+		model.schema.extend( 'listItem', {
+			allowAttributes: [ 'listStyle' ]
+		} );
+
+		// TODO: Is it correct?
+		editor.model.schema.setAttributeProperties( 'listStyle', {
+			isFormatting: true
+		} );
+
+		editor.commands.add( 'listStyles', new ListStylesCommand( editor, DEFAULT_LIST_TYPE ) );
+
+		// Fix list attributes when modifying their nesting levels (the `listIndent` attribute).
+		this.listenTo( editor.commands.get( 'indentList' ), 'executeCleanup', fixListAfterIndentListCommand( editor ) );
+		this.listenTo( editor.commands.get( 'outdentList' ), 'executeCleanup', fixListAfterOutdentListCommand( editor ) );
+
+		this.listenTo( editor.commands.get( 'bulletedList' ), 'executeCleanup', restoreDefaultListStyle( editor ) );
+		this.listenTo( editor.commands.get( 'numberedList' ), 'executeCleanup', restoreDefaultListStyle( editor ) );
+
+		// Register a post-fixer that ensures that the `listStyle` attribute is specified in each `listItem` element.
+		model.document.registerPostFixer( fixListStyleAttributeOnListItemElements( editor ) );
+
+		// Set up conversion.
+		editor.conversion.for( 'upcast' ).add( upcastListItem() );
+		editor.conversion.for( 'downcast' ).add( downcastListStyleAttribute() );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	afterInit() {
+		const editor = this.editor;
+
+		// Enable post-fixer that removes the `listStyle` attribute from to-do list items only if the "TodoList" plugin is on.
+		// We need to registry the hook here since the `TodoList` plugin can be added after the `ListStylesEditing`.
+		if ( editor.commands.get( 'todoList' ) ) {
+			editor.model.document.registerPostFixer( removeListStyleAttributeFromTodoList( editor ) );
+		}
+	}
+}
+
+// Returns a converter that consumes the `style` attribute and search for `list-style-type` definition.
+// If not found, the `"default"` value will be used.
+//
+// @private
+// @returns {Function}
+function upcastListItem() {
```

Comment:
I'd go with a name that reflects upcasting list item's style. The name `upcastListItem()` might suggest that it handles whole `<li>` upcast.